commons/src/main/java/fr/free/nrw/commons/upload/ShareActivity.java

```diff
@@ -168,29 +183,46 @@ public void onCreate(Bundle savedInstanceState) {
             } else {
                 source = Contribution.SOURCE_EXTERNAL;
             }
-
             mimeType = intent.getType();
         }
 
         mediaUriString = mediaUri.toString();
-        Log.d("Image", "Uri: " + mediaUriString);
-
+        Log.d(TAG, "Uri: " + mediaUriString);
         //convert image Uri to file path
         FilePathConverter uriObj = new FilePathConverter(this, mediaUri);
         String filePath = uriObj.getFilePath();
 
+        //Using global singleton to get CacheController to last longer than the activity lifecycle
+        cacheObj = ((CommonsApplication)this.getApplication());
+
         if (filePath != null) {
             //extract the coordinates of image in decimal degrees
-            Log.d("Image", "Calling GPSExtractor");
+            Log.d(TAG, "Calling GPSExtractor");
             GPSExtractor imageObj = new GPSExtractor(filePath);
-            String coords = imageObj.getCoords();
+            String decimalCoords = imageObj.getCoords();
+            double decLongitude = imageObj.getDecLongitude();
+            double decLatitude = imageObj.getDecLatitude();
+
+            if (decimalCoords != null) {
+                Log.d(TAG, "Decimal coords of image: " + decimalCoords);
+                cacheObj.cacheData.setQtPoint(decLongitude, decLatitude);
 
-            if (coords != null) {
-                Log.d("Image", "Coords of image: " + coords);
                 MwVolleyApi apiCall = new MwVolleyApi(this);
 
-                //asynchronous calls to MediaWiki Commons API to match image coords with nearby Commons categories
-                apiCall.request(coords);
+                List displayCatList = cacheObj.cacheData.findCategory();
+
+                //if no categories found in cache, call MW API to match image coords with nearby Commons categories
```

Comment:
This might read better as:

    cacheFound = !displayCatList.isEmpty();
    if (cacheFound) {
        ...
    } else {
      ...
    }